akka-typed/src/main/scala/akka/typed/internal/ActorSystemImpl.scala

```diff
@@ -0,0 +1,251 @@
+
+Streams:
+
+  •  make new implementation of ActorMaterializer that leverages Envelope removal
+  •  all internal actor creation must be asynchronous
+  •  could offer ActorSystem extension for materializer
+  •  remove downcasts to ActorMaterializer in akka-stream package—replace by proper function passing or Materializer APIs where needed (should make Gearpump happier as well)
+  •  add new Sink/Source for ActorRef[]
+
+Distributed Data:
+
+  •  create new Behaviors around the logic
+
+ *
+ */
+
+private[typed] class ActorSystemImpl[-T](
+  override val name: String,
+  _config: Config,
+  _cl: ClassLoader,
+  _ec: Option[ExecutionContext],
+  _p: Props[T])
+    extends ActorRef[T](a.RootActorPath(a.Address("akka", name)) / "user") with ActorSystem[T] with ActorRefImpl[T] {
+
+  import ActorSystemImpl._
+
+  if (!name.matches("""^[a-zA-Z0-9][a-zA-Z0-9-_]*$"""))
+    throw new IllegalArgumentException(
+      "invalid ActorSystem name [" + name +
+        "], must contain only word characters (i.e. [a-zA-Z0-9] plus non-leading '-' or '_')")
+
+  import a.ActorSystem.Settings
+  override val settings: Settings = new Settings(_cl, _config, name)
+
+  override def logConfiguration(): Unit = log.info(settings.toString)
+
+  protected def uncaughtExceptionHandler: Thread.UncaughtExceptionHandler =
+    new Thread.UncaughtExceptionHandler() {
+      def uncaughtException(thread: Thread, cause: Throwable): Unit = {
+        cause match {
+          case NonFatal(_) | _: InterruptedException | _: NotImplementedError | _: ControlThrowable ⇒ log.error(cause, "Uncaught error from thread [{}]", thread.getName)
+          case _ ⇒
+            if (settings.JvmExitOnFatalError) {
+              try {
+                log.error(cause, "Uncaught error from thread [{}] shutting down JVM since 'akka.jvm-exit-on-fatal-error' is enabled", thread.getName)
+                import System.err
+                err.print("Uncaught error from thread [")
+                err.print(thread.getName)
+                err.print("] shutting down JVM since 'akka.jvm-exit-on-fatal-error' is enabled for ActorSystem[")
+                err.print(name)
+                err.println("]")
+                cause.printStackTrace(System.err)
+                System.err.flush()
+              } finally {
+                System.exit(-1)
+              }
+            } else {
+              log.error(cause, "Uncaught fatal error from thread [{}] shutting down ActorSystem [{}]", thread.getName, name)
+              terminate()
+            }
+        }
+      }
+    }
+
+  override val threadFactory: d.MonitorableThreadFactory =
+    d.MonitorableThreadFactory(name, settings.Daemonicity, Option(_cl), uncaughtExceptionHandler)
+
+  override val dynamicAccess: a.DynamicAccess = new a.ReflectiveDynamicAccess(_cl)
+
+  // this provides basic logging (to stdout) until .start() is called below
+  // FIXME!!!
+  private val untypedSystem = a.ActorSystem(name + "-untyped", _config)
```

Comment:
fixing this is a separate ticket—I don’t want to do all the things at once (issues will be created from the PR description shortly)